xarray/tests/test_interp.py

```diff
@@ -662,3 +662,68 @@ def test_datetime_interp_noerror():
         coords={"time": pd.date_range("01-01-2001", periods=50, freq="H")},
     )
     a.interp(x=xi, time=xi.time)  # should not raise an error
+
+
+def interp_1d(data, dim, vals):
+    selectors = {dim: vals}
+
+    old_coord = data[dim]
+
+    lower = data.sel(**selectors, method='ffill')
+    upper = data.sel(**selectors, method='bfill')
+    lower_x = data[dim].sel(**selectors, method='ffill')
+    upper_x = data[dim].sel(**selectors, method='bfill')
+    weight = np.abs(vals - lower_x)/np.abs(upper_x-lower_x)
+    # needed if upper and lower coordinates are identical
+    weight = weight.fillna(1.0)
+    ans = lower * (1-weight) + upper * weight
+    return ans
+
+
+def interp_nd(data, coords):
+    for dim in coords:
+        data = interp_1d(data, dim, coords[dim])
```

Comment:
The downside of this approach (indexing separately along each dimension) is that it is potentially *much* more expensive than doing all indexing at once, e.g., if the result of indexing along multiple dimensions is only a single point.

I think it would be interesting to see if this could be done by indexing all dimensions at once `2**len(coords)` times, with all combinations of bfill/ffill along all dimensions.

This would require modifying `sel` to support a dict of options for `method` indicating different indexing methods along different axes, along the lines of the proposal from https://github.com/pydata/xarray/issues/3223 for `tolerance`.